File: cpp/Array/0069_Sqrt(x).cpp
using namespace std;

//leetcode submit region begin(Prohibit modification and deletion)
class Solution {
public:
  /*
   * binary search
   * searching an array [1, ..., x / 2]
   * */
  int mySqrt(int x) {
    if (x <= 1) return x;
    // long long - because test cases
    long long left = 1, right = x / 2;
    long long mid, square;
    while (left < right) {
      mid = left + ((right - left) >> 1);
      square = mid * mid;
      if (x < square) right = mid;
      else if (x > square) left = mid + 1;
      else return mid;
    }
    return (left * left > x) ? left - 1 : left;
  }
};
//leetcode submit region end(Prohibit modification and deletion)

File: python3/0069-Sqrt(x).py
import math


class Solution:
    def mySqrt(self, x: int) -> int:
        if x <= 1:
            return x

        left, right = 0, x // 2
        while left <= right:
            mid = left + (right - left) // 2
            mid_square = mid * mid
            if mid_square > x:
                right = mid - 1
            else:
                left = mid + 1

        return left if left * left <= x else left - 1


s = Solution()
for i in range(50):
    if s.mySqrt(i) != math.floor(math.sqrt(i)):
        print(i, "false")
